components/ws-proxy/pkg/proxy/infoprovider.go

```diff
@@ -318,9 +344,9 @@ func (c *workspaceInfoCache) reinit(infos []*WorkspaceInfo) {
 	}
 }
 
-func (c *workspaceInfoCache) insert(info *WorkspaceInfo) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (c *workspaceInfoCache) Insert(info *WorkspaceInfo) {
+	c.cond.L.Lock()
+	defer c.cond.L.Unlock()
 
 	c.doInsert(info)
```

Comment:
Why not have the `c.cond.Broadcast()` here instead of in `doInsert`? It would help separate concerns (locking/notification) from insertion.

Reply:
I've moved here - and into `Reinit` - because then
- we don't signal that much during reinit (one broadcast after all insertions)
- the cond.Broadcast() is closer to the locking.